src/limb_base.cpp

```diff
@@ -0,0 +1,126 @@
+/**
+ * @file limb_base.cpp
+ * @author Kodlab - Zac Gong (zacgong@seas.upenn.edu)
+ * @author Kodlab - J. Diego Caporale (jdcap@seas.upenn.edu)
+ * @brief Class implementation for LimbBase.
+ * @date 2022-08-18
+ *
+ * @copyright BSD 3-Clause License, 
+ * Copyright (c) 2021 The Trustees of the University of Pennsylvania. 
+ * All Rights Reserved
+ *
+ */
+
+#include "kodlab_mjbots_sdk/limb_base.h"
+
+namespace kodlab
+{
+    template<typename EndEffectorOutput>
+    LimbBase<EndEffectorOutput>::LimbBase(
+        const std::string &name,
+        const std::vector<std::shared_ptr<JointBase>> &joints,
+        const LimbConfig &config)
+        : joints_(joints), 
+          config_(config),
+          limbDOFs_(joints.size()),
+          name_(name) {}
+
+    template<typename EndEffectorOutput>
+    LimbBase<EndEffectorOutput>::LimbBase(
+        const std::vector<std::shared_ptr<JointBase>> &joints,
+        const LimbConfig &config) 
+        : LimbBase("", joints, config) {}
+
+    template<typename EndEffectorOutput>
+    void LimbBase<EndEffectorOutput>::UpdateJointStates()
+    {
+        if(!joint_positions_.valid()){
+            // update position/velocity/torque vectors
+            joint_positions_.set(joint_utils::PositionsFromJoints(joints_)); 
+        }
+        fk_.invalidate();
+        jac_.invalidate();
+    }
+
+    template<typename EndEffectorOutput>
+    void LimbBase<EndEffectorOutput>::Update()
+    {   
+        UpdateJointStates(); // Get joint states
+
+        // If implemented, fk_ and jac_ are set here and 
+        // the following two lines should do nothing, 
+        // if not function does nothing by default
+        FKAndJacobianImpl(joint_positions_.get()); 
+
+        ForwardKinematics(); 
+        Jacobian();
+    }
+
+    template<typename EndEffectorOutput>
+    EndEffectorOutput LimbBase<EndEffectorOutput>::ForwardKinematics()
+    {
+        if(!fk_.valid()){
+            UpdateJointStates(); 
+            fk_ = ForwardKinematicsImpl(joint_positions_.get());
+        }
+        return fk_;
+    }
+
+    template<typename EndEffectorOutput>
+    Eigen::MatrixXf LimbBase<EndEffectorOutput>::Jacobian()
+    {
+        if(!jac_.valid()){
+            UpdateJointStates();
+            jac_ = JacobianImpl(joint_positions_);
+        }
+        return jac_;
+    }
+
+    template<typename EndEffectorOutput>
+    std::vector<float> LimbBase<EndEffectorOutput>::InverseKinematics(
+                       const EndEffectorOutput &EE_pos)
+    {
+        return InverseKinematicsImpl(EE_pos);
```

Comment:
Should inverse kinematics output be cached?

Reply:
Because the EE_pos is given by the user I didn't think it quite matched the pattern, happy to be overruled.

Reply:
Hmm. I think you are correct. If the user calls IK multiple times with the same input thats on them.

Reply:
Or rather you are I think caching it sounds like a very bad idea.